SteamKit2/SteamKit2/Steam/Handlers/SteamAuthTicket/SteamAuthTicket.cs

```diff
@@ -0,0 +1,249 @@
+            var appTicket = await apps.GetAppOwnershipTicket( appid );
+
+            if ( appTicket.Result != EResult.OK ) throw new Exception( $"Failed to obtain app ownership ticket. Result: {appTicket.Result}. The user may not own the game or there was an error." );
+
+            if ( GameConnectTokens.TryDequeue( out var token ) )
+            {
+                var authTicket = BuildAuthTicket( token );
+                var ticket = await VerifyTicket( appid, authTicket, out var crc );
+
+                // Verify just in case
+                if ( ticket.ActiveTicketsCRC.Any( x => x == crc ) )
+                {
+                    var tok = CombineTickets( authTicket, appTicket.Ticket );
+                    return new TicketInfo( this, appid, tok );
+                }
+                else
+                {
+                    throw new Exception( "Ticket verification failed." );
+                }
+            }
+            else
+            {
+                throw new Exception( "There's no available game connect tokens left." );
+            }
+        }
+
+        internal void CancelAuthTicket( TicketInfo authTicket )
+        {
+            lock ( TicketChangeLock )
+            {
+                if ( TicketsByGame.TryGetValue( authTicket.AppID, out var tickets ) )
+                {
+                    tickets.RemoveAll( x => x.ticket_crc == authTicket.TicketCRC );
+                }
+            }
+
+            SendTickets();
+        }
+
+        private static byte[] CombineTickets( byte[] authTicket, byte[] appTicket )
+        {
+            var len = appTicket.Length;
+            var token = new byte[ authTicket.Length + 4 + len ];
+            var mem = token.AsSpan();
+            authTicket.CopyTo( mem );
+            MemoryMarshal.Write( mem[ authTicket.Length.. ], in len );
+            appTicket.CopyTo( mem[ ( authTicket.Length + 4 ).. ] );
+
+            return token;
+        }
+
+        /// <summary>
+        /// Handles generation of auth ticket.
+        /// </summary>
+        private static byte[] BuildAuthTicket( byte[] gameConnectToken )
+        {
+            const int sessionSize =
+                4 + // unknown, always 1
+                4 + // unknown, always 2
+                4 + // public IP v4, optional
+                4 + // private IP v4, optional
+                4 + // timestamp & uint.MaxValue
+                4;  // sequence
+
+            using var stream = new MemoryStream( gameConnectToken.Length + 4 + sessionSize );
+            using ( var writer = new BinaryWriter( stream ) )
+            {
+                writer.Write( gameConnectToken.Length );
+                writer.Write( gameConnectToken );
+
+                writer.Write( sessionSize );
+                writer.Write( 1 );
+                writer.Write( 2 );
+
+                SerializeRandomUserData( writer );
```

Comment:
```suggestion
                Span<byte> randomBytes = stackalloc byte[8];
                RandomNumberGenerator.Fill( randomBytes );
                writer.Write( randomBytes );
```

that arraypool renting dance is unnecessary, something like this should work.

Reply:
I was basing it off https://github.com/SteamRE/SteamKit/pull/789#issuecomment-899758669 for non-zero bytes implementation, but after testing seems like it works just fine with this suggested change

Reply:
can you apply this and the other comment